aten/src/ATen/MemoryOverlap.cpp

```diff
@@ -34,4 +34,36 @@ void assert_no_internal_overlap(TensorImpl* t) {
     "performing the operation.");
 }
 
+MemOverlapStatus get_overlap_status(const Tensor& a, const Tensor& b) {
+  return get_overlap_status(a.unsafeGetTensorImpl(), b.unsafeGetTensorImpl());
+}
+
+MemOverlapStatus get_overlap_status(TensorImpl* a, TensorImpl* b) {
+  if (a->is_contiguous() && b->is_contiguous() && a->storage().data() == b->storage().data()) {
+    const auto a_begin = static_cast<char*>(a->data());
+    const auto a_end = a_begin + a->numel() * a->itemsize();
+    const auto b_begin = static_cast<char*>(b->data());
+    const auto b_end = b_begin + b->numel() * b->itemsize();
+
+    if (a_begin == b_begin && a_end == b_end) {
+      return MemOverlapStatus::FULL;
+    }
+    if (a_begin < b_end && b_begin < a_end) {
+      return MemOverlapStatus::PARTIAL;
+    }
+  }
+  return MemOverlapStatus::NO;
```

Comment:
This is a little misleading because there could still be memory overlap if a and b are not contiguous but share storage. Example:
```
x = torch.zeros(3, 3)
a = x[0,:]
b = x[:, 0]
# a[2] is shared with b[0]
```
However, in such a case, it is too hard to actually compute if there is memory overlap (this is very similar to the `has_internal_overlap` check). It would be good to either merge the MemOverlap / MemOverlapStatus structs, or to add a TOO_HARD to MemOverlapStatus.